### ISSUE TITLE
[New rule] prefer-async-await

### ISSUE THREAD

**Issue by roman-petrov** [dart-code-checker/dart-code-metrics#664]
During code review today I just noticed a repeated pattern: 

```dart
void processResponse(GetSettingsResponse settings) {
...
}

await SettingsApi().getSettings().then(processResponse);
```

From my knowledge of Dart, there is nothing bad in processing future results using `then` chain. However, my preference and (from my experience) more "standard" way to process promise results would be 

```dart
processResponse(await SettingsApi().getSettings());
```

Personally, I rarely use `then` method to process future results but I am not sure if am alone here :) 

But the reason I created this issue is code style consistency. If there are two ways to do exactly the same thing - we always should choose exactly one way in the same project (without additional reasons).

I know, that if we totally disallow using `then` we might get many annoying warnings in some code bases. But I don't think that this would be too much.

This is not a final rule proposal, just a subject to discuss and maybe we should get some feedback before implementing such rule.

**Comment by incendial** [comment]
Interesting idea, actually I'm not sure, that `await SettingsApi().getSettings().then(processResponse);` and `processResponse(await SettingsApi().getSettings());` works in the same way, since await will also wait for all `then` to complete. Just checked it on dartpad with `Future.delayed` in `then`.

**Comment by incendial** [comment]
But I generally think it's a good check, what do you think about a slightly different name, like `prefer-async-await` or something?

**Comment by roman-petrov** [comment]
Futures and asynchronous code is a complex topic and I think it's out of this issue scope... The goal I am always trying to reach for my codebase is to have simple, straightforward and linear code. IMHO extensive usage of `Future.then` with futures can lead to messy code sometimes. 

> But I generally think it's a good check, what do you think about a slightly different name, like `prefer-async-await` or something?

I think `prefer-async-await` is ok too and maybe it is actually better because it just recommends and there is nothing bad in `Future.then` if used carefully.

**Comment by dkrutskikh** [comment]
Available in 4.11.0 release 🚀
